pkg/common/util/task.go

```diff
@@ -7,46 +7,58 @@ import (
 	"sync"
 )
 
-// RunTasks executes all the provided functions concurrently and waits for
-// them all to complete. If a function returns an error, all other functions
-// are canceled (i.e. the context they are passed is canceled) and the error is
-// returned. If all functions finish to completion successfully, RunTasks
-// returns nil. If the context passed to RunTasks is canceled then each
-// function is canceled and RunTasks returns ctx.Err(). Tasks passed to
-// RunTasks MUST support cancellation via the provided context for RunTasks to
-// work properly.
-func RunTasks(ctx context.Context, tasks ...func(context.Context) error) error {
-	var wg sync.WaitGroup
-	ctx, cancel := context.WithCancel(ctx)
-	defer func() {
-		cancel()
-		wg.Wait()
-	}()
+type TaskRunner struct {
+	wg        sync.WaitGroup
+	ctx       context.Context
```

Comment:
Having a context in a struct is generally discouraged in Go, as it can lead to confusion about context lifetimes and responsibilities: https://go.dev/blog/context-and-structs

From what I can see, the context is stored in the TaskRunner and appears to serve as a global context for all operations. This approach can make it difficult understand when cancellation signals propagate.

Have you considered passing contexts explicitly to methods that need them?

Reply:
May need to discuss fixing the other issue first... I'm thinking we may want ctx here to be able to cancel things on failure...

That said, assuming we don't,
currently it treats all started tasks in the same context. they are all started/canceled in the same context. This makes all tasks run through the same TaskRunner the same.

If we dont do that, we would need to accept a context per set of tasks, and then keep track of all of those contexts in the struct so wait can cancel them all? I figured one context rather then multiple would be simpler/more understandable for this case then multiple.

Reply:
That's fair. Moving the context out of the struct would likely introduce more complexity than it's worth.